doc: using components in Fusion (#258)
vaadin/docs: articles/fusion/application/components.asciidoc

---
title: Using Components
order: 55
layout: page
---
= Using Components in Fusion

> This article describes using Vaadin Components and/or third-party web components in your Fusion views.

Reviewer: ```suggestion
This article describes using Vaadin components and/or third-party web components in your Fusion views.
```
Not a proper product name.

The examples assume a running Fusion application.
For information on setting up a Fusion project and defining a basic view, see the <<../quick-start-tutorial#, Quick Start Tutorial>>.
For a full list of components and detailed API documentation, see the Design System <<{articles}/ds/components#,Components section>>.

> pass:[<!-- vale Vaadin.ProductName = NO -->]

Reviewer 2: The Using Vaadin Components heading is showing incorrectly in the preview, it literally shows
`== Using Vaadin Components`

Author: Done. Changed the title instead of working around the linter.

> When bootstrapping a Fusion project from an https://start.vaadin.com/[official starter], all Vaadin Components frontend packages are automatically installed.

Reviewer: ```suggestion
When bootstrapping a Fusion project from an https://start.vaadin.com/?preset=fusion[official starter], all Vaadin components frontend packages are automatically installed.
```

To use a component in your own view, it is enough to import it into the view `.ts` file.
For example:

[source,typescript]
----
import '@vaadin/vaadin-text-field';
----

> You can then use the `<vaadin-text-field>` element in your view's `render` method:

Reviewer: ```suggestion
You can then use the <<{articles}/ds/components/text-field#,Text Field>> component (the `<vaadin-text-field>` element) in your view's `render` method:
```

[source,typescript]
----
render() {
  return html`
    <vaadin-text-field id='firstname' label='First name'></vaadin-text-field>
    <vaadin-text-field id='lastname' label='Last name'></vaadin-text-field>
  `;
}
----

== Component Class

> The class name of a component `vaadin-_component_` follows the pattern `__Component__Element`

Reviewer: ```suggestion
The class name of a `<vaadin-_component-name_>` element follows the pattern `__ComponentName__Element`
```

Importing the class gives a type safe access to the component's API.
For example:

[source, typescript]
----
include::{root}/frontend/demo/fusion/components/greeting-view.ts[]
----

> In the above example, the combo box and checkbox components with ids `greeting` and `custom` respectively are mapped to typed class variables of the same names via the `@query` decorator.

Reviewer: ```suggestion
In the above example, the <<{articles}/ds/components/combo-box#,Combo Box>> and <<{articles}/ds/components/checkbox#,Checkbox>> components with IDs `greeting` and `custom` respectively are mapped to typed class variables of the same names via the `@query` decorator.
```

> The checkbox event listener toggles the `allowCustomValue` attribute of the combo box, which defines whether the user can enter custom values into the box or is constrained to the alternatives in the dropdown.

Reviewer: ```suggestion
The Checkbox event listener toggles the `allowCustomValue` attribute of the Combo Box, which defines whether the user can enter custom values into the field or is constrained to the alternatives in the items overlay.
```

Having strongly typed class variables helps to catch errors at compile time and allows IDE completion.

== HTML Attributes

> The following HTML attributes work as expected with all Vaadin Components:

Reviewer: ```suggestion
The following HTML attributes work as expected with all Vaadin components:
```

> 

Reviewer: Give more space to the description.
```suggestion
[cols=1,2, opts=header]
```

The following HTML attributes work as expected with all Vaadin components:
[cols=1,2, opts=header]
|===
|Property | Description

> 

Reviewer 2: The table looks misformatted in the preview.
![Screenshot 2021-03-15 at 10 38 43](https://user-images.githubusercontent.com/30408303/111125768-a541f680-857a-11eb-96e1-7158c8bee07d.png)

Author: Done.

|`title`
|Additional label associated with the component, shown as a tooltip.
|`visible`
|Visibility; set to `false` to hide the element.
|===

See the <<{articles}/ds/components#,Components section>> for component-specific attributes and methods.

> There you can also find information on how to customize the look and feel of the components.

Reviewer: ```suggestion
The <<{articles}/ds/customization#, Customization>> section describes how to customize the look and feel of the components.
```

== Using a Third-Party Component

You can also install and use other components in Fusion views.

> To do this, you first need to install the component npm package from `npmjs.com` using the `pnpm` package manager.

Reviewer: ```suggestion
To do this, you first need to install the component npm package from the https://npmjs.com[npm package registry] using the `pnpm` package manager.
```

For example, this command installs the latest version of the `vanilla-colorful` color picker and records the dependency in `package.json`:

`npx pnpm add vanilla-colorful`

After installation, the component is now ready to be imported and used in the Fusion view.

[NOTE]
If your project is configured to use the `npm` package manager instead of `pnpm`, use the command `npm install vanilla-colorful --save`.

> The following example shows the hex triplet of the selected color in a `vaadin-text-field`:

Reviewer: Prefer using proper component names instead of class/element names.

```suggestion
The following example shows the hex triplet of the selected color in a <<{articles}/ds/components/text-field#,Text Field>>:
```

The following example shows the hex triplet of the selected color in a <<{articles}/ds/components/text-field#,Text Field>>:

[source,typescript]
----
include::{root}/frontend/demo/fusion/components/color-picker-view.ts[]

> ----

Reviewer 2: Is it possible to have the code example to show the live demo as well? as what @platosha  did in his [PR](https://github.com/vaadin/docs/pull/240). It has the benefit of being able to see the code and demo at the same time. https://vaadin.com/docs-beta/latest/fusion/application/events/.
Does @jouni have some tips on how to make that kind of include?

Author: Made both examples live using `render` tag in the `include`. Would there be some way to make the code appear before the live demo (as I feel in this context the code is more important than the running example itself).

Reviewer 2: could change [source, typescript] to [source, typescript, role=show-code], so it will always show the code without having to click the "show code".

Author: Thanks for the tip, fixed.

Reviewer 2: just one more small thing, I learned from Jouni that the `group=TypeScript` is not needed unless we have the same example with both TypeScript and Java examples.

Author: Done.
